### DISCUSSION TITLE
Proposal: Allow integration specific automation triggers to be shown in the visual automation editor

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#795] [General] [answered]
There has been a lot of work done recently to make the visual automation editor more powerful. Each trigger platform has it's own set of UI components which gets us one step closer to allowing users to build complex automations completely through the UI.

Some time ago, we introduced the concept of integrations providing their own triggers, and `zwave_js` has taken advantage of that to provide useful device triggers. Our documentation also includes information on how to use these triggers directly via YAML, but there is currently no mechanism to support these in the UI. This proposal is to add a mechanism that would allow integrations to provide the information needed to build a UI for these custom triggers:

Solution:
1. Add support for a `triggers.yaml` file where an integration can describe each trigger and provide the necessary selectors. Like `services.yaml`, the core can pass this information to the frontend so that the UI can be dynamically constructed. The one downside to this approach is that while the UI side might be dynamic, the content side is static. It might be more useful to have something like a config flow where inputs can dictate the outputs and options can be determined dynamically.
2. Figure out how to register custom triggers with the core. Right now there is no registration, it's all done dynamically, but the core would need to know which custom triggers were valid in order to present them in the automation UI. One way to do this is to figure it out when we load in the integration instead of having a separate registration mechanism (like with services) but open to suggestions here.
3. Update the frontend to query a websocket API when loading the visual editor that returns all of the relevant custom triggers and selector information so that the UI can be extended to include these integrations as platforms.

This needs to be a little better thought out, but it seems like a natural evolution of the work we are doing with the UI already so I wanted to put this out there and start getting feedback.

**Comment by Jc2k** [comment]
Sorry if this is a dumb thing to ask and there is some difference I am overlooking, but I thought that's what device automations let us do. Eg for the homekit controller devices there are custom automation triggers (like double tap on a doorbell), and by defining device automations I can trigger a automation when someone using my someone triggers that trigger entirely in the UI. A lot of what you describe (including APIs for enumerating the triggers and the UI that use them) are already built for device automations.

**Reply by raman325** [reply]
you're right for the most part. In the case of zwave_js, we have two custom triggers, one that triggers when a value on the device changes, the other that triggers when an event comes in from the controller. For the value updates, that's straight forward and we have it all worked out as a device trigger. For events, we could instrument every single event but it seems counterproductive. With that being said a user may have a use case to automate a certain event, so the device trigger wouldn't work for them

**Comment by thecode** [comment]
Although my answer is not fully aligned with the discussion, I think the support for an integration to add custom triggers should be implemented into core.

I have copied the logic implemented in `zwave_js` for use in `webostv` integration and my code is very similar to the copied one.
https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave_js/trigger.py
https://github.com/home-assistant/core/tree/dev/homeassistant/components/zwave_js/triggers

https://github.com/home-assistant/core/blob/dev/homeassistant/components/webostv/trigger.py
https://github.com/home-assistant/core/tree/dev/homeassistant/components/webostv/triggers

If an integration wants to implemented a custom trigger, the developer will need again to copy this, assuming he found these examples, which currently only exists in these two integrations.

I think similar to device automations we should have a platform for custom triggers, so that core will handle the logic to register a new trigger and the integration will only need to implement the trigger itself. This will also align with the proposal topic here, since once it is standardized in core, we can also make sure the visual editor will be able to support it.

**Comment by frenck** [comment, marked as answer]
Quick update, this is actually being worked on by the team at the Open Home Foundation right now.

Hold on a little longer (should be in the upcoming releases as of writing this message.

I'll go ahead and mark this as "accepted" from an architectural proposal, and close it for now.

../Frenck
